mappings/ctv3sctmap2/import_data.py

```diff
@@ -1,45 +1,83 @@
 import csv
-import datetime
 import glob
 import os
+import sqlite3
 
-from django.db import transaction
-
-from .models import Mapping
+from django.db import connection as django_connection
 
 
 def import_data(release_dir):
-    def load_records():
-        paths = glob.glob(
-            os.path.join(
-                release_dir,
-                "Mapping Tables",
-                "Updated",
-                "Clinically Assured",
-                "ctv3sctmap2*.txt",
-            )
+    """
+    Import NHSD CTV3 -> SNOMED concept maps
+    """
+    paths = glob.glob(
+        os.path.join(
+            release_dir,
+            "Mapping Tables",
+            "Updated",
+            "Clinically Assured",
+            "ctv3sctmap2*.txt",
         )
-        assert len(paths) == 1
-
-        with open(paths[0]) as f:
-            reader = csv.reader(f, delimiter="\t")
-            next(reader)
-            for r in reader:
-                if r[4] == "_DRUG":  # SCT_CONCEPTID
-                    r[4] = None
-                    r[5] = None
-                r[6] = r[6] == "1"  # MAPSTATUS
-                r[7] = parse_date(r[7])  # EFFECTIVEDATE
-                r[8] = r[8] == "1"  # IS_ASSURED
-
-                if not r[6]:
-                    continue
-                yield r
-
-    with transaction.atomic():
-        Mapping.objects.all().delete()
-        Mapping.objects.bulk_create(Mapping(*r) for r in load_records())
-
-
-def parse_date(datestr):
-    return datetime.date(int(datestr[:4]), int(datestr[4:6]), int(datestr[6:]))
+    )
+    assert len(paths) == 1
+
+    with open(paths[0]) as f:
+        reader = csv.DictReader(f, delimiter="\t")
+
+        values = list(iter_values(reader))
+
+    # UPSERT rows based on ID, using effective date to decide if a row should
+    # overwrite an existing one.
+    query = """
+    INSERT INTO ctv3sctmap2_mapping(
+        id,
+        ctv3_concept_id,
+        ctv3_term_id,
+        ctv3_term_type,
+        sct_concept_id,
+        sct_description_id,
+        map_status,
+        effective_date,
+        is_assured)
+      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
+    ON CONFLICT(id) DO UPDATE SET
+      map_status=excluded.map_status,
+      effective_date=excluded.effective_date,
+      is_assured=excluded.is_assured
+    WHERE excluded.effective_date > ctv3sctmap2_mapping.effective_date;
+    """
+
+    # execute the query above for each row from the release data
+    connection_params = django_connection.get_connection_params()
+    connection = sqlite3.connect(**connection_params)
+    connection.executemany(query, values)
+    connection.commit()
+    connection.close()
+
+
+def iter_values(rows):
+    """
+    Convert the given rows into an iterable of values for upserting to the DB.
+    """
+    for r in rows:
+        # ignore SNOMED Concept and Description IDs when the Concept ID is _DRUG
+        sct_concept = r["SCT_CONCEPTID"] if r["SCT_CONCEPTID"] != "_DRUG" else ""
```

Comment:
For drugs, this has changed from `None` to `""` -- is this deliberate?

Reply:
Nope, good catch!

Reply:
Fixed in fe08eb4